assets/js/modules/analytics-4/utils/data-mock.test.js

```diff
@@ -0,0 +1,195 @@
+				endDate: 'not-a-date',
+				startDate: '2020-12-31',
+			} )
+		).toThrow( 'a valid endDate is required' );
+	} );
+
+	it( 'generates a valid report with no compare dates', () => {
+		const report = getAnalytics4MockResponse( {
+			startDate: '2020-12-31',
+			endDate: '2021-01-27',
+			metrics: [
+				{
+					name: 'sessions',
+				},
+				{
+					name: 'newUsers',
+				},
+			],
+			dimensions: [ 'date' ],
+		} );
+
+		const firstRow = report[ 0 ].rows[ 0 ];
+		expect( firstRow ).toHaveProperty( 'dimensionValues' );
+		expect( firstRow ).toHaveProperty( 'metricValues' );
+	} );
+
+	it( 'generates the correct number of rows when there is no date range', () => {
+		const report = getAnalytics4MockResponse( {
+			startDate: '2020-12-01',
+			endDate: '2020-12-05',
+			metrics: [
+				{
+					name: 'totalUsers',
+				},
+				{
+					name: 'averageSessionDuration',
+				},
+			],
+			dimensions: [ 'date' ],
+		} );
+
+		expect( report[ 0 ].rows ).toHaveLength( 5 );
+	} );
+
+	it( 'generates the correct number of rows when there is a date range', () => {
+		const report = getAnalytics4MockResponse( {
+			startDate: '2020-12-01',
+			endDate: '2020-12-05',
+			compareStartDate: '2020-11-26',
+			compareEndDate: '2020-11-30',
+			metrics: [
+				{
+					name: 'totalUsers',
+				},
+				{
+					name: 'averageSessionDuration',
+				},
+			],
+			dimensions: [ 'date' ],
+		} );
+
+		expect( report[ 0 ].rows ).toHaveLength( 20 );
+	} );
+
+	it( 'checks the number of metric values generated', () => {
+		const report = getAnalytics4MockResponse( {
+			startDate: '2020-12-01',
+			endDate: '2020-12-15',
+			compareStartDate: '2020-11-26',
+			compareEndDate: '2020-11-30',
+			metrics: [
+				{
+					name: 'totalUsers',
+				},
+				{
+					name: 'averageSessionDuration',
+				},
+				'sessions',
+			],
+			dimensions: [ 'date' ],
+		} );
+
+		expect( report[ 0 ].rows[ 0 ].metricValues ).toHaveLength( 3 );
+	} );
+
+	it( 'checks the number of dimension values generated', () => {
+		const report = getAnalytics4MockResponse( {
+			startDate: '2020-12-01',
+			endDate: '2020-12-15',
+			metrics: [ 'sessions' ],
+			dimensions: [ 'date', { name: 'country' } ],
+		} );
+
+		expect( report[ 0 ].rows[ 0 ].dimensionValues ).toHaveLength( 2 );
+	} );
+
+	it( 'checks if aggregations are rendered', () => {
+		const report = getAnalytics4MockResponse( {
+			startDate: '2020-12-01',
+			endDate: '2020-12-15',
+			metrics: [ 'sessions' ],
+			dimensions: [ 'date' ],
+		} );
+
+		expect( report[ 0 ] ).toHaveProperty( 'totals' );
+		expect( report[ 0 ] ).toHaveProperty( 'minimums' );
+	} );
+
+	it( 'checks if the report structure matches what we get from the response', () => {
+		const report = getAnalytics4MockResponse( {
+			startDate: '2022-11-01',
+			endDate: '2022-12-31',
+			metrics: [ 'totalUsers', 'averageSessionDuration', 'sessions' ],
+			dimensions: [ 'date', 'country' ],
+		} );
+
+		expect( report[ 0 ].dimensionHeaders ).toMatchObject(
```

Comment:
@techanvil I don't think we can use `toEqual` to check between the response and the data generated since the data generated is dynamic and will never match what we have in the JSON file. That's why I compared the length of some elements.
I've also added test to verify that the number of date_range_0 rows matches the number of date_range_1 rows

Reply:
Hey @asvinb, actually, as described below the RNG utility has an explicit seed set so the generated values are the same for each combination of report args. I gave it a quick test by snapshotting the report and it does work. So, we should be able to use `toEqual` here.

https://github.com/google/site-kit-wp/blob/0f06107080a26ad10d8b855bdcc8261472841331/assets/js/modules/analytics-4/utils/data-mock.js#L226-L228

Reply:
@techanvil I've updated the PR, can you take another look please?
Thanks!